### ISSUE TITLE
PulseAudio driver (or other variable chunk size driver) + LADSPA plugins

### ISSUE THREAD

**Issue by theGreatWhiteShark** [hydrogen-music/hydrogen#1381]
Hydrogen version *          :
Operating system + version  : Linux
Audio driver + version      :  PulseAudio

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

While reading #1378 I wondered how the LADSPA plugins do respond to dynamic buffer sizing. As this is also done by the PulseAudio driver I just tried it and the results are really bad. At least with my machine the audible artifacts are beyond being tolerable.

As there is still another open issue with PulseAudio I will have a look at the API soon. Maybe Hydrogen could bargain for a lower bound in the buffer sizes. I will check. If not, we should at least put a warning icon next to the FX section in the Mixer which explains the matter using a tooltip.

**Comment by cme** [comment]
Oh, I didn't realise PulseAudio had variable buffer sizing too. Any idea what causes the artefacts -- is it underruns due to extra processing time, or plain old bugs where a fixed buffer size is assumed?

**Comment by theGreatWhiteShark** [comment]
> Oh, I didn't realise PulseAudio had variable buffer sizing too.

Haha, actually you were the one pointing me towards this :wink: 

> Any idea what causes the artefacts -- is it underruns due to extra processing time, or plain old bugs where a fixed buffer size is assumed?

Hard to tell. The lowest buffer size I experienced was 1. To get a feeling of what is happening, I measured the average time consumed by 4 LADSPA plugins and the total processing time with respect to the buffer size used in `audioEngine_process`.

```
Ladspa process time:  0.969467  Total process time:  1.22353  Buffer size (process):  135  Buffer size (driver):  511
Ladspa process time:  0.776067  Total process time:  1.0124  Buffer size (process):  240  Buffer size (driver):  511
Ladspa process time:  0.692333  Total process time:  0.8942  Buffer size (process):  127  Buffer size (driver):  511
Ladspa process time:  0.5626  Total process time:  0.734667  Buffer size (process):  97  Buffer size (driver):  511
Ladspa process time:  0.643467  Total process time:  0.77  Buffer size (process):  151  Buffer size (driver):  511
Ladspa process time:  0.680333  Total process time:  0.757133  Buffer size (process):  128  Buffer size (driver):  511
Ladspa process time:  0.8708  Total process time:  0.9508  Buffer size (process):  127  Buffer size (driver):  511
Ladspa process time:  0.9068  Total process time:  0.994933  Buffer size (process):  15  Buffer size (driver):  511
Ladspa process time:  0.764533  Total process time:  0.831867  Buffer size (process):  151  Buffer size (driver):  511
Ladspa process time:  1.0256  Total process time:  1.116  Buffer size (process):  240  Buffer size (driver):  511
Ladspa process time:  0.978533  Total process time:  1.07013  Buffer size (process):  143  Buffer size (driver):  511
Ladspa process time:  1.1278  Total process time:  1.2266  Buffer size (process):  248  Buffer size (driver):  511
Ladspa process time:  1.47367  Total process time:  1.5988  Buffer size (process):  256  Buffer size (driver):  511
Ladspa process time:  1.03253  Total process time:  1.12447  Buffer size (process):  29  Buffer size (driver):  511
Ladspa process time:  0.501867  Total process time:  0.552933  Buffer size (process):  166  Buffer size (driver):  511
Ladspa process time:  0.438533  Total process time:  0.487533  Buffer size (process):  38  Buffer size (driver):  511
Ladspa process time:  0.324267  Total process time:  0.3584  Buffer size (process):  29  Buffer size (driver):  511
Ladspa process time:  0.333867  Total process time:  0.3738  Buffer size (process):  68  Buffer size (driver):  511
Ladspa process time:  0.265  Total process time:  0.297267  Buffer size (process):  62  Buffer size (driver):  511
Ladspa process time:  0.240733  Total process time:  0.27  Buffer size (process):  53  Buffer size (driver):  511
Ladspa process time:  0.288067  Total process time:  0.325333  Buffer size (process):  37  Buffer size (driver):  511
Ladspa process time:  0.2648  Total process time:  0.293133  Buffer size (process):  22  Buffer size (driver):  511
Ladspa process time:  0.307933  Total process time:  0.349667  Buffer size (process):  23  Buffer size (driver):  511
Ladspa process time:  0.2578  Total process time:  0.289467  Buffer size (process):  40  Buffer size (driver):  511
Ladspa process time:  0.224533  Total process time:  0.248067  Buffer size (process):  6  Buffer size (driver):  511
Ladspa process time:  0.4432  Total process time:  0.4896  Buffer size (process):  38  Buffer size (driver):  511
Ladspa process time:  0.532867  Total process time:  0.584267  Buffer size (process):  81  Buffer size (driver):  511
Ladspa process time:  1.4024  Total process time:  1.51447  Buffer size (process):  256  Buffer size (driver):  511
```

511 was the buffer size I selected via the Preferences.

I haven't looked into the LADSPA and am not sure whether it would be a bug to assume a fixed buffer size in a plugin. After all, the target environment is Linux + JACK driver. The latter supports only fixed buffer size to the power of two, which makes a lot of sense FFT-wise. But I tested for PortAudio whether it would make a difference to have a buffer size of 512 or 511 and it does not. (Although, I'm not sure whether the tested plugins do really work in frequency space or not). Anyway, with PortAudio various sizes, even very small ones, performed nicely.

**Comment by cme** [comment]
I *think* LADSPA plugins must be able to handle arbitrary buffer sizes, since it's a parameter to every `run()` call which implies a requirement that it can be anything, but I can't find anything that states that explicitly.

**Comment by theGreatWhiteShark** [comment]
Let's see. I have to toy around a little bit more to grasp what's going on.

I recently read that some of the [CAPS LADSPA](http://quitte.de/dsp/caps.html#Operation) plugins may not work with sample rates other than 44100 and 48000, which I found quite surprising since a lot of professional stuff happens at 92k. Who knows what other people have baked into their plugins.

**Comment by cme** [comment]
I tried some LADSPA plugins in a Linux VM and I *think* I'm hearing some artefacts even with JACK. I can't tell if this is just because it's a VM, though.

*Some* of the plugins I've tried do have much worse sound with PulseAudio than LADSPA, so I'll take a look at the sources of those to see what's going on. If some plugins have requirements like fixed size buffers or specific sample rates, I guess we could add some (optional) resampling and buffer chunking, which would be a pain because it'd increase latency, but if plugins requirements are only a *fixed* size buffer, we can allow that fixed size to be very small

**Comment by theGreatWhiteShark** [comment]
> I tried some LADSPA plugins in a Linux VM and I think I'm hearing some artefacts even with JACK. I can't tell if this is just because it's a VM, though.

Hmm. Using JACK everything sound fine at my end.

> If some plugins have requirements like fixed size buffers or specific sample rates, I guess we could add some (optional) resampling and buffer chunking, which would be a pain because it'd increase latency, but if plugins requirements are only a fixed size buffer, we can allow that fixed size to be very small

I bet the LADSPA API doesn't feature communicating such constrains. Also, I haven't sufficiently looked into PulseAudio to understand what is going wrong. In JACK AFAIK there is a known timeout for how long the audio is allowed to be processed before reaching the system's sink. If an app takes too long, there will be a XRUN. Is this true for PulseAudio as well (with variable timeouts due to dynamic buffer sizes)? Or does the server waits and does not guarantee real time processing? Since the audio output of Hydrogen itself sound legit, I would tend except the second option.

**Comment by cme** [comment]
Okay, found the source of the problem. And a couple of others to boot. 

`renderNoteResample` has cut and paste code for calculating the output sample values. It does it once for calculating the instrument output to mixer, advances the buffer index into the sample, and then recalculates the output values to feed to each plugin. This has the effect that the plugin inputs are always slightly ahead of the main output, by nFrames worth. If nFrames changes between callbacks, the position of the plugin input will jump by the difference, giving discontinuities in the audio data.

The other effect this has is that the plugins never actually see the first nFrames of any instrument's sample.

Another thing I've found is that the ADSR envulope is applied to the main output, but is *not* applied to the plugin send data. This seems wrong.

I'll fix these, probably by refactoring a bit to remowe the duplication, and probably also to avoid calculating sample values multiple times, since that and calculating the adsr coefficient for each sample are things i remember seeing prominently in profiling data at nome point in the past, probably on raspberry pi.

Another issue i noticed is that interpolation has a rough edge around nFrames boundaries, and could do better by holding over some state between callbacks. This currently causes some clicks during loud sample playback.

**Comment by cme** [comment]
Actually, I was wrong about the interpolation issue. I misread the code, and thought that was accounting for the audible things I was hearing, which on reflection *must* just be because it's in a VM, because they'd be audible on all platforms if this was the case! Excellent.

**Comment by theGreatWhiteShark** [comment]
Puh. Sounds like the state of the audio engine is more dire than I thought.

> Another thing I've found is that the ADSR envulope is applied to the main output, but is not applied to the plugin send data. This seems wrong.

Yes. That's a bug too as the plugins do not operate on the actual audio of the note. Just confirmed it.

**Comment by cme** [comment]
Great, that makes fixing and refactoring simpler :)

**Comment by cme** [comment]
Other bugs in this code found by inspection:

  - using any of the multi-sample interpolation methods, one sample past the beginning of the buffer will be read. This doesn't have much effect, but it shouldn't be read.
  - The per-instrument "low pass resonant" filter (a 2-tap IIR filter), if active, will end with the sample's end. In practice this probably has almost no effect, but in cases were a sample ends abruptly with a discontinuity, the filter won't be able to smooth out any 'click' at the end.
    - is this filter ever actually used though?

**Comment by theGreatWhiteShark** [comment]
> is this filter ever actually used though?

I doubt it. I use tons of EQ to remove resonances in order to make the kits sound nice. But why just a low pass and not even a low and high pass combination?

If at some point I'll be really bored and have finished all my books, I might overhaul the FX section of the mixer. But this has very little priority. Until then I would keep them because if a user has activated them on more than four instruments she can not reproduce the same setup with the four aux send FX in the mixer via a LADSPA plugin.

**Comment by cme** [comment]
I've addressed all those in #1403 plus another couple of rough edges around the end of samples that  almost certainly don't show up in practice, even if they did wouldn't be noticable, and even if they did they wouldn't be noticed. But they were there so I fixed them 😆
